# Pandas


Pandas is a library built using NumPy specifically for data analysis. You'll be using Pandas heavily for data manipulation, visualisation, building machine learning models, etc. 

There are two main data structures in Pandas - Series and Dataframes. The default way to store data is dataframes, and thus manipulating dataframes quickly is probably the most important skill set for data analysis. 

<img src="pandas_1.png">

*source: https://pandas.pydata.org/pandas-docs/stable/overview.html*

In this section, you will study:
1. The pandas Series (similar to a numpy array)
    * Creating a pandas series
    * Indexing series
2. Dataframes 
    * Creating dataframes from dictionaries
    * Importing CSV data files as pandas dataframes
    * Reading and summarising dataframes
    * Sorting dataframes 

### 1. The Pandas Series 

A series is similar to a 1-D numpy array, and contains scalar values of the same type (numeric, character, datetime etc.). 
A dataframe is simply a table where each column is a pandas series.


#### Creating Pandas Series

Series are one-dimensional array-like structures, though unlike numpy arrays, they often contain non-numeric data (characters, dates, time, booleans etc.)

You can create pandas series from array-like objects using ```pd.Series()```.

In [1]:
# import pandas, pd is an alias
import pandas as pd

# Creating a numeric pandas series
s = pd.Series([2, 4, 5, 6, 9])
print(s)
print(type(s))

0    2
1    4
2    5
3    6
4    9
dtype: int64
<class 'pandas.core.series.Series'>


Note that each element in the Series has an index, and the index starts at 0 as usual.

In [2]:
# creating a series of characters
# notice that the 'dtype' here is 'object'
char_series = pd.Series(['a', 'b', 'af'])
char_series

0     a
1     b
2    af
dtype: object

In [3]:
# creating a series of type datetime
date_series = pd.date_range(start = '11-09-2017', end = '12-12-2017')
date_series
type(date_series)


pandas.core.indexes.datetimes.DatetimeIndex

#### Indexing Series

Indexing series is exactly same as 1-D numpy arrays - index starts at 0.

In [4]:
# Indexing pandas series: Same as indexing 1-d numpy arrays or lists
# accessing the fourth element
s[3]

# accessing elements starting index = 2 till the end
s[2:]

2    5
3    6
4    9
dtype: int64

In [5]:
# accessing the second and the fourth elements
# note that s[1, 3] will not work, you need to pass the indices [1, 3] as a list inside the original []
s[[1, 3]]

1    4
3    6
dtype: int64

Usually, you will work with Series only as a part of dataframes. Let's study the basics of dataframes.

### The Pandas Dataframe

Dataframe is the most widely used data-structure in data analysis. It is a table with rows and columns, with rows having an index and columns having meaningful names.

#### Creating dataframes from dictionaries

There are various ways of creating dataframes, such as creating them from dictionaries, JSON objects, reading from txt, CSV files, etc. 

In [6]:
# keys become column names
df = pd.DataFrame({'name': ['Vinay', 'Kushal', 'Aman', 'Saif'], 
                   'age': [22, 25, 24, 28], 
                    'occupation': ['engineer', 'doctor', 'data analyst', 'teacher']})
df

age    name    occupation
0   22   Vinay      engineer
1   25  Kushal        doctor
2   24    Aman  data analyst
3   28    Saif       teacher

#### Importing CSV data files as pandas dataframes 

For the upcoming exercises, we will use a dataset of a retail store having details about the orders placed, customers, product details, sales, profits etc. 



In [7]:
# reading a CSV file as a dataframe
market_df = pd.read_csv("../global_sales_data/market_fact.csv")


Usually, dataframes are imported as CSV files, but sometimes it is more convenient to convert dictionaries 
into dataframes. For e.g. when the raw data is in a JSON format (which is not uncommon), you can easily convert it into a dictionary, and then into a dataframe. 

You will learn how to convert JSON objects to dataframes later.

#### Reading and Summarising Dataframes

After you import a dataframe, you'd want to quickly understand its structure, shape, meanings of rows and columns etc. Further, you may want to look at summary statistics - such as mean, percentiles etc.

In [8]:
# Looking at the top and bottom entries of dataframes
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

In [9]:
market_df.tail()

Ord_id  Prod_id   Ship_id    Cust_id      Sales  Discount  \
8394  Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08   
8395  Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10   
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798   243.0500      0.02   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03   
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798   603.6900      0.00   

      Order_Quantity  Profit  Shipping_Cost  Product_Base_Margin  
8394              28  374.63           7.69                 0.59  
8395              20  -74.03           6.92                 0.37  
8396              39  -70.85           5.35                 0.40  
8397              23  565.34          30.00                 0.62  
8398              47  131.39           4.86                 0.38

Here, each row represents an order placed at a retail store. Notice the index associated with each row - starts at 0 and ends at 8398, implying that there were 8399 orders placed.

In [10]:
# Looking at the datatypes of each column
market_df.info()

# Note that each column is basically a pandas Series of length 8399
# The ID columns are 'objects', i.e. they are being read as characters
# The rest are numeric (floats or int)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8399 entries, 0 to 8398
Data columns (total 10 columns):
Ord_id                 8399 non-null object
Prod_id                8399 non-null object
Ship_id                8399 non-null object
Cust_id                8399 non-null object
Sales                  8399 non-null float64
Discount               8399 non-null float64
Order_Quantity         8399 non-null int64
Profit                 8399 non-null float64
Shipping_Cost          8399 non-null float64
Product_Base_Margin    8336 non-null float64
dtypes: float64(5), int64(1), object(4)
memory usage: 656.2+ KB


In [11]:
# Describe gives you a summary of all the numeric columns in the dataset
market_df.describe()

Sales     Discount  Order_Quantity        Profit  Shipping_Cost  \
count   8399.000000  8399.000000     8399.000000   8399.000000    8399.000000   
mean    1775.878179     0.049671       25.571735    181.184424      12.838557   
std     3585.050525     0.031823       14.481071   1196.653371      17.264052   
min        2.240000     0.000000        1.000000 -14140.700000       0.490000   
25%      143.195000     0.020000       13.000000    -83.315000       3.300000   
50%      449.420000     0.050000       26.000000     -1.500000       6.070000   
75%     1709.320000     0.080000       38.000000    162.750000      13.990000   
max    89061.050000     0.250000       50.000000  27220.690000     164.730000   

       Product_Base_Margin  
count          8336.000000  
mean              0.512513  
std               0.135589  
min               0.350000  
25%               0.380000  
50%               0.520000  
75%               0.590000  
max               0.850000

In [12]:
# Column names
market_df.columns

Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Sales', 'Discount',
       'Order_Quantity', 'Profit', 'Shipping_Cost', 'Product_Base_Margin'],
      dtype='object')

In [13]:
# The number of rows and columns
market_df.shape

(8399, 10)

In [14]:
# You can extract the values of a dataframe as a numpy array using df.values 
market_df.values

array([['Ord_5446', 'Prod_16', 'SHP_7609', ..., -30.51, 3.6, 0.56],
       ['Ord_5406', 'Prod_13', 'SHP_7549', ..., 4.56, 0.93, 0.54],
       ['Ord_5446', 'Prod_4', 'SHP_7610', ..., 1148.9, 2.5, 0.59],
       ..., 
       ['Ord_5388', 'Prod_6', 'SHP_7524', ..., -70.85, 5.35, 0.4],
       ['Ord_5348', 'Prod_15', 'SHP_7469', ..., 565.34, 30.0, 0.62],
       ['Ord_5459', 'Prod_6', 'SHP_7628', ..., 131.39, 4.86, 0.38]], dtype=object)

#### Indices 

An important concept in pandas dataframes is that of *row indices*. By default, each row is assigned indices starting from 0, and are represented at the left side of the dataframe. 

In [15]:
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

Now, arbitrary numeric indices are difficult to read and work with. Thus, you may want to change the indices of the df to something more meanigful.

Let's change the index to Ord_id (unique id of each order), so that you can select rows using the order ids directly.

In [16]:
# Setting index to Ord_id
market_df.set_index('Ord_id', inplace = True)
market_df.head()

Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                      
Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446   -30.51           3.60                 0.56  
Ord_5406     4.56           0.93                 0.54  
Ord_5446  1148.90           2.50                 0.59  
Ord_5456   729.34          14.30                 0.37  
Ord_5485  1219.87          26.30                 0.38

Having meaningful row labels as indices helps you to select (subset) dataframes easily. You will study selecting dataframes in the next section. 

#### Sorting dataframes

You can sort dataframes in two ways - 1) by the indices and 2) by the values.  


In [17]:
# Sorting by index
# axis = 0 indicates that you want to sort rows (use axis=1 for columns)
market_df.sort_index(axis = 0, ascending = False)

Prod_id   Ship_id   Cust_id       Sales  Discount  Order_Quantity  \
Ord_id                                                                        
Ord_999   Prod_15  SHP_1383  Cust_361   5661.0800      0.00              33   
Ord_998    Prod_8  SHP_1380  Cust_372    750.6600      0.00              33   
Ord_998    Prod_5  SHP_1382  Cust_372   2149.3700      0.03              42   
Ord_998    Prod_8  SHP_1381  Cust_372    254.3200      0.01               8   
Ord_997   Prod_14  SHP_1379  Cust_365  28761.5200      0.04               8   
Ord_996    Prod_7  SHP_1378  Cust_371     35.6400      0.05              10   
Ord_996    Prod_5  SHP_1377  Cust_371   3202.2500      0.09              44   
Ord_996    Prod_6  SHP_1378  Cust_371     81.2500      0.01              11   
Ord_996   Prod_13  SHP_1378  Cust_371     57.1700      0.08              18   
Ord_996   Prod_13  SHP_1378  Cust_371     41.9700      0.05              12   
Ord_995    Prod_4  SHP_1376  Cust_370   7965.9025      0.01              46   
Ord_994    Prod_1  SHP_1375  Cust_369  11103.7500      0.10              32   
Ord_993    Prod_1  SHP_1374  Cust_368   2645.8000      0.00              31   
Ord_993    Prod_5  SHP_1373  Cust_368   1140.2600      0.04              39   
Ord_992   Prod_15  SHP_1370  Cust_368   5041.4600      0.06              43   
Ord_992   Prod_11  SHP_1372  Cust_368   8817.7100      0.09              27   
Ord_992    Prod_6  SHP_1371  Cust_368    282.9800      0.06              45   
Ord_991    Prod_4  SHP_1369  Cust_367    760.2485      0.04              16   
Ord_990   Prod_11  SHP_1368  Cust_366   1042.0600      0.04              14   
Ord_990    Prod_6  SHP_1367  Cust_366    492.7100      0.02              24   
Ord_99     Prod_8   SHP_136   Cust_72    311.4400      0.04              32   
Ord_99     Prod_5   SHP_137   Cust_72    215.6500      0.01              17   
Ord_989    Prod_3  SHP_1366  Cust_360     64.1300      0.10              10   
Ord_988    Prod_7  SHP_1364  Cust_360     90.9600      0.02              48   
Ord_988    Prod_5  SHP_1364  Cust_360    632.5500      0.05              27   
Ord_988    Prod_5  SHP_1365  Cust_360   1997.1300      0.05              32   
Ord_987    Prod_6  SHP_1362  Cust_365    905.3200      0.02              24   
Ord_987    Prod_4  SHP_1363  Cust_365   1702.4990      0.08              16   
Ord_986    Prod_3  SHP_1360  Cust_363     77.5700      0.00              19   
Ord_985    Prod_4  SHP_1359  Cust_362   1155.5835      0.04              24   
...           ...       ...       ...         ...       ...             ...   
Ord_1016   Prod_8  SHP_1406  Cust_383    138.0500      0.09              19   
Ord_1015   Prod_9  SHP_1405  Cust_370    124.0600      0.09              17   
Ord_1014   Prod_5  SHP_1404  Cust_368    131.0900      0.01              10   
Ord_1014   Prod_6  SHP_1404  Cust_368   1817.9000      0.06              33   
Ord_1013   Prod_4  SHP_1403  Cust_382   1064.7865      0.10              12   
Ord_1012   Prod_3  SHP_1402  Cust_382    104.5100      0.04              13   
Ord_1011   Prod_3  SHP_1400  Cust_381   3465.9700      0.08               4   
Ord_1011   Prod_8  SHP_1401  Cust_381   1307.8800      0.06              13   
Ord_1010  Prod_11  SHP_1399  Cust_366   4051.8000      0.10              35   
Ord_1010   Prod_5  SHP_1398  Cust_366    195.2300      0.08              35   
Ord_101    Prod_1   SHP_139   Cust_73    300.2000      0.00               8   
Ord_101    Prod_4   SHP_139   Cust_73   1364.8025      0.07              29   
Ord_1009  Prod_17  SHP_1396  Cust_366   2219.7900      0.10              30   
Ord_1009   Prod_6  SHP_1397  Cust_366    237.3600      0.07              25   
Ord_1008   Prod_5  SHP_1395  Cust_380    689.6700      0.07              38   
Ord_1007   Prod_4  SHP_1394  Cust_379   2766.8180      0.03              47   
Ord_1006   Prod_2  SHP_1393  Cust_378   1758.4100      0.06              18   
Ord_1005  Prod_13  SHP_1392  Cust_377     35.4800      0.09 

In [18]:
# Sorting by values

# Sorting in increasing order of Sales
market_df.sort_values(by='Sales').head()

Prod_id   Ship_id    Cust_id  Sales  Discount  Order_Quantity  \
Ord_id                                                                    
Ord_704    Prod_7   SHP_964   Cust_242   2.24      0.01               1   
Ord_149    Prod_3  SHP_7028  Cust_1712   3.20      0.09               1   
Ord_4270   Prod_7  SHP_5959  Cust_1450   3.23      0.06               2   
Ord_4755  Prod_13  SHP_6628  Cust_1579   3.41      0.06               1   
Ord_2252   Prod_3  SHP_3064   Cust_881   3.42      0.05               1   

          Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                
Ord_704    -1.97           0.70                 0.37  
Ord_149    -3.16           1.49                 0.37  
Ord_4270   -2.73           0.70                 0.81  
Ord_4755   -1.78           0.70                 0.56  
Ord_2252   -2.91           1.49                 0.37

In [19]:
# Sorting in decreasing order of Shipping_Cost
market_df.sort_values(by='Shipping_Cost', ascending = False).head()

Prod_id   Ship_id   Cust_id      Sales  Discount  Order_Quantity  \
Ord_id                                                                       
Ord_1751  Prod_15  SHP_2426  Cust_597  14740.510      0.00              46   
Ord_839   Prod_11  SHP_1361  Cust_364  12689.870      0.04              44   
Ord_1741  Prod_11  SHP_2411  Cust_595  15168.820      0.02              26   
Ord_417   Prod_11   SHP_561  Cust_156  20333.816      0.02              45   
Ord_1581  Prod_15  SHP_2184  Cust_519   2573.920      0.07              17   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_1751  3407.73         164.73                 0.56  
Ord_839   -169.23         154.12                 0.76  
Ord_1741 -1096.78         147.12                 0.80  
Ord_417  -1430.45         147.12                 0.80  
Ord_1581   117.23         143.71                 0.55

In [20]:
# Sorting by more than two columns

# Sorting in ascending order of Sales for each Product
market_df.sort_values(by=['Prod_id', 'Sales'], ascending = False)

Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                     
Ord_2197  Prod_9  SHP_2994   Cust_827  7522.80      0.04              48   
Ord_4356  Prod_9  SHP_6074  Cust_1481  6831.72      0.01              41   
Ord_262   Prod_9   SHP_358    Cust_66  6553.45      0.03              39   
Ord_4059  Prod_9  SHP_5660  Cust_1378  5587.20      0.05              36   
Ord_2973  Prod_9  SHP_6073  Cust_1480  5410.95      0.09              36   
Ord_950   Prod_9  SHP_1315   Cust_334  4906.85      0.09              32   
Ord_5112  Prod_9  SHP_7141  Cust_1729  4273.95      0.05              49   
Ord_612   Prod_9   SHP_836   Cust_687  3872.38      0.10              50   
Ord_3443  Prod_9  SHP_4773  Cust_1246  3849.17      0.06              46   
Ord_3650  Prod_9  SHP_6795  Cust_1683  3353.54      0.07              22   
Ord_4590  Prod_9  SHP_6385  Cust_1465  3212.97      0.03              37   
Ord_2774  Prod_9  SHP_3809  Cust_1015  3191.24      0.00              38   
Ord_793   Prod_9  SHP_1089   Cust_245  2833.19      0.06              34   
Ord_3071  Prod_9  SHP_4265  Cust_1141  2728.42      0.06              45   
Ord_1290  Prod_9  SHP_1780   Cust_482  2609.53      0.00              27   
Ord_555   Prod_9   SHP_754   Cust_202  2568.71      0.10              43   
Ord_2853  Prod_9  SHP_3927  Cust_1056  2366.18      0.10              16   
Ord_4370  Prod_9  SHP_6092  Cust_1492  2325.42      0.02              25   
Ord_3239  Prod_9  SHP_4491  Cust_1205  2300.45      0.02              36   
Ord_2968  Prod_9  SHP_4092  Cust_1066  2160.83      0.08              36   
Ord_4989  Prod_9  SHP_6965  Cust_1700  2048.78      0.05              24   
Ord_4663  Prod_9  SHP_6498  Cust_1612  2026.42      0.03              50   
Ord_1152  Prod_9  SHP_1586   Cust_443  1554.53      0.00              16   
Ord_5370  Prod_9  SHP_7500  Cust_1809  1527.42      0.04              25   
Ord_3917  Prod_9  SHP_5439  Cust_1337  1504.01      0.05              42   
Ord_4595  Prod_9  SHP_6390  Cust_1523  1500.84      0.10              46   
Ord_3389  Prod_9  SHP_4698  Cust_1236  1484.89      0.03              41   
Ord_604   Prod_9   SHP_826   Cust_213  1406.64      0.02              50   
Ord_2737  Prod_9  SHP_3755   Cust_994  1400.91      0.09              49   
Ord_2461  Prod_9  SHP_3373   Cust_941  1396.75      0.04              46   
...          ...       ...        ...      ...       ...             ...   
Ord_1666  Prod_1  SHP_2303   Cust_490   101.52      0.02               6   
Ord_3693  Prod_1  SHP_5118  Cust_1287    96.03      0.10               6   
Ord_2133  Prod_1  SHP_2910   Cust_808    96.01      0.01              11   
Ord_3579  Prod_1  SHP_4956  Cust_1272    94.99      0.06              13   
Ord_1654  Prod_1  SHP_2284   Cust_491    90.58      0.06               7   
Ord_5479  Prod_1  SHP_7655  Cust_1805    88.94      0.05               5   
Ord_3506  Prod_1  SHP_4857   Cust_395    79.93      0.00               2   
Ord_1858  Prod_1  SHP_2820   Cust_720    73.86      0.08               7   
Ord_75    Prod_1    SHP_99    Cust_50    73.07      0.08               4   
Ord_5107  Prod_1  SHP_7132  Cust_1708    72.72      0.06               4   
Ord_2725  Prod_1  SHP_3736   Cust_997    70.02      0.00               4   
Ord_1933  Prod_1  SHP_2649   Cust_707    69.57      0.05               8   
Ord_4613  Prod_1  SHP_6420  Cust_1569    68.03      0.02               4   
Ord_3379  Prod_1  SHP_4685  Cust_1233    66.92      0.01               5   
Ord_542   Prod_1   SHP_735   Cust_188    63.84      0.02               3   
Ord_210   Prod_1   SHP_288    Cust_94    63.52      0.08               4   
Ord_4618  Prod_1  SHP_6432  Cust_1575    58.90      0.03               5   
Ord_212   Prod_1   SHP_291    Cust_88    58.33      0.02               3   
Ord_2410  Prod_1  SHP_3307   Cust_963    57.48      0.00               3   
Ord_4213  Prod_1  SHP_5875  Cust_1428    53.30      0.00 